components/backup_text/src/mask.rs

```diff
@@ -0,0 +1,195 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::convert::{TryFrom, TryInto};
+
+use tidb_query_datatype::codec::{
+    data_type::{DateTime, Duration, Enum},
+    Datum,
+};
+
+use crate::eval_context;
+
+pub type MaskResult = Result<Datum, Datum>;
+
+const DEFAULT_CONTEXT: &'static str = "tidb";
+
+#[inline]
+fn new_hasher() -> blake3::Hasher {
+    blake3::Hasher::new_derive_key(DEFAULT_CONTEXT)
+}
+
+fn hash_bytes(bytes: &[u8], size: usize) -> Vec<u8> {
+    let mut hasher = new_hasher();
+    hasher.update(bytes);
+    let mut reader = hasher.finalize_xof();
+
+    let mut buf = vec![0; size];
+    reader.fill(&mut buf);
+    buf
+}
+
+const I24_MIN: i64 = -(1 << 24 - 1);
+const I24_MAX: i64 = (1 << 24 - 1) - 1;
+const U24_MIN: u64 = 0;
+const U24_MAX: u64 = (1 << 24) - 1;
+
+#[inline]
+fn mask_i64(from: i64) -> i64 {
+    let to = i64::from_le_bytes(hash_bytes(&mut from.to_le_bytes(), 8).try_into().unwrap());
+
+    if let Ok(_) = i8::try_from(from) {
+        to % (i8::MAX as i64 + 1)
+    } else if let Ok(_) = i16::try_from(from) {
+        to % (i16::MAX as i64 + 1)
+    } else if let I24_MIN..=I24_MAX = from {
+        to % (I24_MAX as i64 + 1)
+    } else if let Ok(_) = i32::try_from(from) {
+        to % (i32::MAX as i64 + 1)
+    } else {
+        to
+    }
+}
+
+#[inline]
+fn mask_u64(from: u64) -> u64 {
+    let to = u64::from_le_bytes(hash_bytes(&mut from.to_le_bytes(), 8).try_into().unwrap());
+
+    if let Ok(_) = u8::try_from(from) {
+        to % (u8::MAX as u64 + 1)
+    } else if let Ok(_) = u16::try_from(from) {
+        to % (u16::MAX as u64 + 1)
+    } else if let U24_MIN..=U24_MAX = from {
+        to % (U24_MAX as u64 + 1)
+    } else if let Ok(_) = u32::try_from(from) {
+        to % (u32::MAX as u64 + 1)
+    } else {
+        to
+    }
+}
+
+#[inline]
+fn mask_f64(f: f64) -> f64 {
+    f64::from_le_bytes(hash_bytes(&mut f.to_le_bytes(), 8).try_into().unwrap())
+}
+
+fn mask_string(bytes: &[u8]) -> Vec<u8> {
+    let size = bytes.len();
+    let sum = hash_bytes(bytes, size / 2);
+    let mut hex = hex::encode(sum);
+    hex.push_str(&"*".repeat(size - hex.len()));
+    hex.into_bytes()
+}
+
+fn mask_duration(dur: Duration) -> Duration {
+    let secs = dur.to_secs() / 3600 * 3600;
```

Comment:
This may not "mask" enough, we still keep the `hours` info. Or if all of the user's data is concentrated in some scope (like less than an hour), all the masked duration will become the same value